### ISSUE TITLE
Performance gap with `tf.data.Dataset` on MNIST

### ISSUE THREAD

**Issue by rademacher-p** [google/grain#1164]
I'm experiencing a serious performance gap between `tf.data.Dataset` using TFRecord and `grain` w/ ArrayRecord. Simply trying to iterate over MNIST with comparable speed. Snippet below exemplifies, Grain seems 5-10x slower.

Trying to make comparison fair, so I've disabled caching and prefetch for TF. I've tried adding a variety of Grain functionality to improve speed, including `mp_prefetch`, varying `num_threads`, using `CacheIterDataset` and `ThreadPrefetchIterDataset`, etc.

FYI, I've also found one TF option that slows the performance down to Grain-level: `num_parallel_calls_for_decode` in `tfds.ReadConfig`.

```python
import time
from pathlib import Path

import grain.python as grain
import tensorflow as tf
import tensorflow_datasets as tfds
import tqdm

tf.config.set_visible_devices([], device_type="GPU")


n_epoch = 10
batch_size = 32


def make_tf_dataset():
    ds = tfds.load(
        "mnist",
        split="train",
        data_dir=Path.home() / "datasets/tfrecord",
        read_config=tfds.ReadConfig(
            options=tf.data.Options(), try_autocache=False, skip_prefetch=True
        ),
    )

    ds = ds.repeat(n_epoch)
    ds = ds.batch(batch_size)
    ds = tfds.as_numpy(ds)

    return ds


def make_grain_dataset():
    source = tfds.data_source(
        "mnist", split="train", data_dir=Path.home() / "datasets/array_record"
    )
    ds = grain.MapDataset.source(source)

    ds = ds.repeat(n_epoch)
    ds = ds.batch(batch_size)
    ds = ds.to_iter_dataset(grain.ReadOptions(num_threads=8))

    return ds


def profile(ds):
    t0 = time.perf_counter()
    for _ in tqdm.tqdm(iter(ds)):
        pass
    t1 = time.perf_counter()
    return t1 - t0


if __name__ == "__main__":
    t_tf = profile(make_tf_dataset())
    print(f"TF: {t_tf:.3f} s")

    t_grain = profile(make_grain_dataset())
    print(f"Grain: {t_grain:.3f} s")
```

I keep trying to make Grain work for me every few months (since #569), but with no avail, and it seems like other performance issues are going largely unresolved. I love JAX and want to switch to Grain, but trying to get it to work is making me ☹️

**Comment by iindyk** [comment]
Sorry to hear you're experiencing performance issues. Can you please try out the performance debug mode: https://google-grain.readthedocs.io/en/latest/tutorials/dataset_debugging_tutorial.html#debug-mode and share the output? CProfile would also be helpful

From your observation I think the source will be the bottleneck, and, if you try to dig further, most likely image decoding part in TFDS is the bottleneck. If your actual training setup is planning to use imagenet data, we can help you profile and optimize these further, but since this functionality itself lives in TFDS there's not much grain can do, we'd have to optimize those.

fwiw, we have an e2e imagenet pipeline that's on par with tf.data (and we're planning to externalize it), but it has all the necessary transformations: the ones you have + shuffle, shard, random resize crops, random flips. We're also continuously working on perf improvements (the current ones include batching pushdown to the source, prefetch injection and in longer term a C++ backend).

**Comment by rademacher-p** [comment]
@iindyk Thanks for the response - enjoyed your recent DevLab talk! I ran my code with `py_debug_mode=True` (and the code from the debug mode docs page), but didn't see anything on stdout... Can raise a separate issue if that's helpful. Note: Python 3.13, `grain==0.2.15`, and `jax==0.8.0`

As you predicted, cProfile seems to confirm that the bottleneck is related to image decoding. However, there is significant time spent in the array record reader, as well. If I skip `tfds.data_source` and use `ArrayRecordDataSource` directly, I still see time spent in `__getitem__`, but no decoding delay (since the source returns bytes?).

Would be very interested to see an example of ArrayRecord sources being loaded at `tf.data` speed. Would also like to see some documented examples for small datasets and `InMemoryDataSource`. 🙏 

Using `tf.data.Dataset`:
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    18760   18.377    0.001   18.621    0.001 {built-in method tensorflow.python._pywrap_tfe.TFE_Py_FastPathExecute}
  204/201    1.293    0.006    1.314    0.007 {built-in method _imp.create_dynamic}
     4809    1.058    0.000    1.058    0.000 {built-in method marshal.loads}
    18751    0.487    0.000   24.168    0.001 iterator_ops.py:756(_next_internal)
    37543    0.438    0.000    0.444    0.000 {built-in method tensorflow.python.util._pywrap_utils.FlattenForData}
```

Using Grain w/ `tfds.data_source`:
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
600000/4164   93.665    0.000    0.652    0.000 {imdecode}
599998/39726   87.866    0.000    7.736    0.000 array_record_data_source.py:341(__getitem__)
  1200659   12.044    0.000   12.044    0.000 {built-in method numpy.array}
600000/599995   11.870    0.000   48.934    0.000 example_parser.py:138(_features_to_numpy)
  1200000    8.686    0.000   22.706    0.000 example_parser.py:185(_feature_to_numpy)
1200002/1199992    8.046    0.000   13.834    0.000 features_dict.py:248(_flatten)
600000/4164    6.602    0.000    0.820    0.000 image_feature.py:357(decode_example_np_with_opencv)
   600000    5.856    0.000    5.856    0.000 {method 'repeat' of 'numpy.ndarray' objects}
 513794/0    5.659    0.000    0.000          source.py:80(__getitem__)
599999/4161    5.080    0.000    1.062    0.000 top_level_feature.py:58(_decode_example_generic)
   600000    4.726    0.000    8.756    0.000 features_dict.py:272(_nest)
   600000    4.587    0.000    4.587    0.000 {built-in method FromString}
37500/16578    4.549    0.000    2.188    0.000 shape_base.py:380(stack)
  1837592    4.227    0.000    5.219    0.000 contextlib.py:108(__init__)
1199999/8322    4.169    0.000    0.951    0.000 top_level_feature.py:186(_decode_feature)
600000/599995    3.900    0.000   62.147    0.000 example_parser.py:130(parse_example)
3678056/3678054    3.711    0.000    4.746    0.000 {built-in method builtins.next}
   600000    3.479    0.000   12.845    0.000 image_feature.py:614(_reshape_grayscale_image)
```

**Comment by iindyk** [comment]
glad you attended!
thanks for sharing the profile. both imdecode and source __getitem__ should benefit from increased parallelism, unless you're running on a machine with 8 cpus? perhaps GIL contention from other ops slows it down when parallelism is increased

other noticeable things: built-in method numpy.array is making a data copy which I think tfds does when parsing the data, _features_to_numpy also does a copy because of bytes and tf.Example protobuf memory layout in C++ and Python. We have an internal parser that avoids a data copy from example to numpy -- we should release that externally.

TFE_Py_FastPathExecute in tf.data path just lumps all of the execution, I wish it showed more details of where the time is spent

**Comment by innat** [comment]
@rademacher-p 

Could you please let me know the reason behind you set `tf.config.set_visible_devices([], device_type="GPU")`? Recently I tried to use `tf` APIs as processing function inside the grain dataloader and setting `num_threads > 1` causes some issue. The error kinda disappear if we disable tf to reach GPU. https://github.com/tensorflow/tensorflow/issues/123980

**Comment by rademacher-p** [comment]
@innat I've been training JAX models, so it's just been my habit to disable TF from accessing my GPUs. Don't want TF as a dependency for my project at all, which is why I was really hoping `grain` would liberate me from needing `tensorflow_datasets`
